tsinfer/eval_util.py

```diff
@@ -716,10 +716,141 @@ def count_sample_child_edges(ts):
     """
     child_counts = np.bincount(ts.tables.edges.child)
     return child_counts[ts.samples()]
-# def mean_sample_ancestry(ts):
-#     """
-#     Computes the mean sample ancestry for each node in the tree sequence. This is
-#     defined as the fraction of samples below a given node from each population,
-#     averaged along the length of sequence that the node is ancestral to at
-#     least one sample.
-#     """
+
+
+def node_span(ts):
+    """
+    Returns the "span" of all nodes in the tree sequence. This is defined as the
+    total distance along the sequence of all trees that the node participates in.
+    A node "participates in" a tree if it is reachable from a root. The span of all
+    samples is therefore equal to the sequence length.
+    """
+    S = np.zeros(ts.num_nodes)
+    start = np.zeros(ts.num_nodes) - 1
+    iterator = zip(ts.edge_diffs(), ts.trees(sample_counts=False))
+    for ((left, _), edges_out, edges_in), tree in iterator:
+        for edge in edges_out:
+            u = edge.parent
+            # This is complicated a bit by having to support trees in which we
+            # can have branches leading from a root which have zero samples.
+            for u in [edge.parent, edge.child]:
+                condition = (
+                        tree.parent(u) == msprime.NULL_NODE and
+                        tree.left_child(u) == msprime.NULL_NODE and
+                        not ts.node(u).is_sample() and
+                        start[u] != -1)
+                if condition:
+                    S[u] += left - start[u]
+                    start[u] = -1
+        for edge in edges_in:
+            for u in [edge.parent, edge.child]:
+                if start[u] == -1:
+                    start[u] = left
+    for u in tree.nodes():
+        S[u] += ts.sequence_length - start[u]
+    return S
+
+
+def mean_sample_ancestry(ts, sample_sets, show_progress=False):
```

Comment:
Do we want to call this "sample_ancestry" or simply something like "mean_sample_location"? We want to be careful that to avoid terminology that implies anything about the actual location/details of this ancestral node